### PR TITLE
Merge rustc into rmc - 2021 week 38 (#501)

### LINKED ISSUE #501: 2021 week 38: Merge changes from rustc repository
We currently develop RMC on a fork of the rust compiler. We synchronize with the upstream project once a week. This issue is to track the merge for week 38.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Weekly synchronization between RMC and its upstream rustc repo. The last commit is for implementing ShallowBoxInit which was recently introduced to rustc.

### Resolved issues:

Resolves #501 and #528

### Call-outs:

Since this is a merge PR, please *do not merge squash* it. Once it is approved I'll push it via command line.

### Testing:

* How is this change tested? 
Regression

* Is this a refactor change?
No

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by celinval** [comment]
1. Please ignore the format-check since this is checking for copyright header on rustc files.
2. Please review https://github.com/model-checking/rmc/pull/530/commits/fdd9d9e92970c6e1d13cfc056ed6cc9e2a35eaaf

**Review by tedinski — APPROVED** [review]
I can't really speak to whether it's right, but it looks plausible to me, and I think we'd rather merge and find the issue later than block this PR for too long.

**Comment by danielsn** [comment]
LGTM modulo our discussion about documentation

**Pull request closed without merging**
